rllib/offline/estimators/tests/test_ope.py

```diff
@@ -199,22 +217,440 @@ def test_fqe_model(self):
         tmp_probs = []
         for act in range(fqe.policy.action_space.n):
             tmp_actions = np.zeros_like(self.batch["actions"]) + act
-            log_probs = fqe.policy.compute_log_likelihoods(
+            log_probs = self.policy.compute_log_likelihoods(
                 actions=tmp_actions,
                 obs_batch=self.batch["obs"],
             )
-            tmp_probs.append(torch.exp(log_probs))
-        tmp_probs = torch.stack(tmp_probs).transpose(0, 1)
-        tmp_probs = convert_to_numpy(tmp_probs)
+            tmp_probs.append(np.exp(log_probs))
+        tmp_probs = np.stack(tmp_probs).T
         check(action_probs, tmp_probs, decimals=3)
 
-    def test_multiple_inputs(self):
-        # TODO (Rohan138): Test with multiple input files
-        pass
+    def test_fqe_optimal_convergence(self):
+        # Optimal CliffWalkingWallPolicy with epsilon = 0.0
+        # and CliffWalkingWallEnv are deterministic;
+        # check that FQE converges to the true Q-values for self.batch
+        q_vals = [
+            -2.50,
+            -1.51,
+            -0.52,
+            0.49,
+            1.50,
+            2.53,
+            3.56,
+            4.61,
+            5.67,
+            6.73,
+            7.81,
+            8.90,
+            10.00,
+        ]
+        q_model_config = {
+            "tau": 1.0,
+            "model": {
+                "fcnet_hiddens": [],
+                "activation": "linear",
+            },
+            "lr": 0.01,
+            "n_iters": 5000,
+            "delta": 1e-3,
+        }
+
+        fqe = FQETorchModel(
+            policy=self.policy,
+            gamma=self.gamma,
+            **q_model_config,
+        )
+        losses = fqe.train(self.batch)
+        print(losses[-10:])
+        assert losses[-1] < fqe.delta, "FQE loss did not converge!"
+        estimates = fqe.estimate_v(self.batch)
+        print(estimates)
+        check(
+            estimates,
+            q_vals,
+            decimals=1,
+        )
+
+
+def get_cliff_walking_wall_policy_and_data(
+    num_episodes: int,
+    gamma: float,
+    epsilon: float,
+) -> Tuple[Policy, SampleBatch, float, float]:
+    """Collect a cliff_walking_wall policy and data with epsilon-greedy exploration.
+
+    Args:
+        num_episodes: Number of episodes to collect
```

Comment:
You don't exactly collect num_episodes episodes. The logic rnow collects at least num_episodes episodes. if that difference is not important please update the docstring. otherwise make sure you break out of the loop as soon as num_episodes episodes are collected.

Reply:
**TODO**